src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -750,20 +651,32 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 		bool queryOK = false;
 		bool failOnError = false;
 		int64 currentAffectedTupleCount = 0;
-		PGconn *connection = GetConnectionForPlacement(taskPlacement,
-													   isModificationQuery);
+		int connectionFlags = 0;
+		MultiConnection *connection;
 
-		if (connection == NULL)
+		if (isModificationQuery)
+		{
+			connectionFlags |= FOR_DML;
+		}
+
+		/*
+		 * FIXME: It's not actually correct to use only one shard placement
+		 * here for router queries involving multiple relations. We should
+		 * check that this connection is the only modifying one associated
+		 * with all the involved shards.
+		 */
```

Comment:
Any thoughts on how this will work?

It might be useful to have something like `GetPlacementJoinConnection(connectionFlags, taskPlacement, task->relationShardList)`, that looks up the placementEntry for all shards in relationShardList and returns a modifying connection for any of the placements or errors out if there is more than one.

On the other hand, If we allow CTEs that modify multiple shards then this API might not be sufficient.